test/quicapitest.c

```diff
@@ -2345,6 +2345,129 @@ static int test_early_ticks(void)
+    SSL *qlistener = NULL, *qserver = NULL, *qconn = 0;
+    static unsigned char alpn[] = { 8, 'o', 's', 's', 'l', 't', 'e', 's', 't' };
+    int testresult = 0;
+    int chk;
+    BIO *lbio = NULL, *sbio = NULL;
+    BIO_ADDR *addr = NULL;
+    struct in_addr ina;
+    int bio_caps = BIO_DGRAM_CAP_HANDLES_DST_ADDR | BIO_DGRAM_CAP_HANDLES_SRC_ADDR;
+
+    ina.s_addr = htonl(0x1f000001);
+    if (!TEST_ptr(lctx = SSL_CTX_new_ex(libctx, NULL, OSSL_QUIC_server_method()))
+        || !TEST_ptr(sctx = SSL_CTX_new_ex(libctx, NULL, OSSL_QUIC_server_method()))
+        || !TEST_true(SSL_CTX_use_certificate_file(lctx, cert, SSL_FILETYPE_PEM))
+        || !TEST_true(SSL_CTX_use_PrivateKey_file(lctx, privkey, SSL_FILETYPE_PEM))
+        || !TEST_true(SSL_CTX_use_certificate_file(sctx, cert, SSL_FILETYPE_PEM))
+        || !TEST_true(SSL_CTX_use_PrivateKey_file(sctx, privkey, SSL_FILETYPE_PEM))
+        || !TEST_true(BIO_new_bio_dgram_pair(&lbio, 0, &sbio, 0)))
+        goto err;
+
+    if (!TEST_true(BIO_dgram_set_caps(lbio, bio_caps))
+        || !TEST_true(BIO_dgram_set_caps(sbio, bio_caps)))
+        goto err;
+
+    if (!TEST_ptr(addr = BIO_ADDR_new()))
+        goto err;
+
+    if (!TEST_true(BIO_ADDR_rawmake(addr, AF_INET, &ina, sizeof(ina),
+                                    htons(8040))))
+        goto err;
+
+    if (!TEST_int_eq(BIO_dgram_set0_local_addr(lbio, addr), 1))
+        goto err;
+    addr = NULL;
+
+    if (!TEST_ptr(addr = BIO_ADDR_new()))
+        goto err;
+
+    if (!TEST_true(BIO_ADDR_rawmake(addr, AF_INET, &ina, sizeof(ina),
+                                    htons(4080))))
+        goto err;
+
+    if (!TEST_int_eq(BIO_dgram_set0_local_addr(sbio, addr), 1))
+        goto err;
+    addr = NULL;
+
+    SSL_CTX_set_alpn_select_cb(lctx, select_alpn, NULL);
+    SSL_CTX_set_verify(lctx, SSL_VERIFY_NONE, NULL);
+    SSL_CTX_set_alpn_select_cb(sctx, select_alpn, NULL);
+    SSL_CTX_set_verify(sctx, SSL_VERIFY_NONE, NULL);
+
+    if (!TEST_ptr(qserver = SSL_new_listener(lctx, 0)))
+        goto err;
+
+    SSL_set_bio(qserver, sbio, sbio);
+    sbio = NULL;
+
+    if (!TEST_true(SSL_listen(qserver)))
+        goto err;
+
+    if (!TEST_ptr(qlistener = SSL_new_listener(lctx, 0)))
+        goto err;
+
+    SSL_set_bio(qlistener, lbio, lbio);
+    lbio = NULL;
+
+    if (!TEST_true(SSL_listen(qlistener)))
+        goto err;
+
+    if (!TEST_ptr(qconn = SSL_new_from_listener(qlistener, 0)))
+        goto err;
+
+    if (!TEST_false(SSL_set_alpn_protos(qconn, alpn, sizeof(alpn))))
+        goto err;
+
+    if (!TEST_ptr(addr = BIO_ADDR_new()))
+        goto err;
+
+    if (!TEST_true(BIO_ADDR_rawmake(addr, AF_INET, &ina, sizeof(ina),
+                                    htons(4080))))
+        goto err;
+
+    if (!TEST_true(SSL_set1_initial_peer_addr(qconn, addr)))
+        goto err;
+
+    while ((chk = SSL_do_handshake(qconn)) == -1) {
+        SSL_handle_events(qserver);
+        SSL_handle_events(qlistener);
+    }
```

Comment:
Again this sort of thing could be a helper function in the style of `qtest_create_quic_connection` but using a real server rather than a tserver.